### ISSUE TITLE
JetEngine v3.6.0 Beta

### ISSUE THREAD

**Issue by MjHead** [CrocoBlock/suggestions#7842]
Hi,

We are almost getting ready to release JetEngine v3.6.0 and we would like to ask you to help us test it.

You can download beta version from your Crocoblock account.

Please note that is a beta version, not stable release, so its not ready to use on production websites!

## AI Website Structure Builder

This functionality enables the creation of a data model for a website based on user input. The model includes:

- CPTs (Custom Post Types);
- Taxonomies for these CPTs;
- Meta fields (with or without Custom Meta Storage);
- CCTs (Custom Content Types) for entities that don’t require separate front-end pages;
- Basic queries and listings for CPTs and CCTs. A basic query targets the correct type with the selected CPT or CCT as an argument. A basic listing shows items with all fields for a CPT/CCT without styling;
- Relations;
- Filters.

### Logic and UI

<img width="1461" alt="image" src="https://github.com/user-attachments/assets/9e70e8aa-52fb-41cf-836f-602767321b11">

For setup, users describe their desired website in 1–2 sentences. The first field covers the site’s general purpose, while the second field outlines specific functionalities, such as listings, filters, and visitor capabilities. Once both fields are completed, the “Preview Website Model” button is enabled. Clicking it generates a data model for review without actually creating it on the site. Users can review and adjust the model before moving to the creation step.

To create all the entities outlined in the model, check the box labeled “I understand that by clicking ‘Create Website Model,’ all described entities will be physically created on my website,” then click “Create Website Model.”
After this, all entities from the model are created, and the user is redirected to a page with details about the created model.

![image](https://github.com/user-attachments/assets/f1e7664e-8619-49ce-ad00-51d916dfefe8)

Now, users can edit entities or related elements, rename the model (useful when managing multiple models), check the tutorials for individual entities, or the recommended tutorials for further steps.

The final component of the builder is a list of all previously created models.

![image](https://github.com/user-attachments/assets/158495ea-ae33-40e1-8bd6-9ff934fbfde3)

When hovering over a model, a control appears to delete it. This removes only the model information but not the entities it created. Full entity deletion isn’t currently available, as it poses risks. However, based on demand, we may add this feature. 

![image](https://github.com/user-attachments/assets/2d961b9b-dd45-47e1-9bc9-4ca45bd3a83d)

The created models page and the individual model page serve as a central hub for site administrators, consolidating controls for managing the entire data structure in one place. This setup also provides an overall view of each entity’s purpose on the site and its relationships with other entities.

## JetEngine Calendar Module. Min/Max Year and Selects for Years and Months

Now, you can set a `min` and `max` year in the Calendar JetEngine module.

Previously, the only way to switch to a different month in the calendar was by using arrows. This was not convenient if you needed to go many years back or forth, e.g., from  November 2024 to March 1992, as it required a lot of clicks.

### Enable Date Selection

To solve this, a new setting called Allow date select was added to the Dynamic Calendar. 

![image](https://github.com/user-attachments/assets/26b56343-0bcd-47b8-8586-e5ea7a64a3b2)

When enabled, the month and year in the calendar header become clickable. 

![image](https://github.com/user-attachments/assets/a4c9b09a-5f50-4deb-95db-df5af8882455)

By clicking, a dropdown select menu appears, allowing you to choose the desired month or year. Once selected, it navigates immediately, so going from November 2024 to March 1992 now only requires selecting March and then 1992, instead of almost many clicks on arrows. 

### Additional Settings

![image](https://github.com/user-attachments/assets/97f0c046-ac8d-429d-8ca1-6a3170ad1731)
- **Min select year** - the minimum year available in the dropdown for selecting a year. If the “Hide past events” option is enabled, the minimum year automatically matches the current year.
- **Max select year** - the maximum year available in the dropdown. 

These controls allow you to limit the year selection to only the years present in the query, which cannot be auto-detected and must be estimated by the user. 

By default (if max and min years are not set here, and the controls are left empty), the available date range will be 1970-2038.

### Setting Min and Max Years Depending on the Current Year

Sometimes, you may want to set min and max years relative to the current year, such as [current year (2024) minus 5 years]. This can be done as follows:

- Min select year: –5years (2024 – 5 = 2019);
- Max select year:  this year (current year, 2024);

The resulting dropdown would look like this:

![image](https://github.com/user-attachments/assets/640ada3b-bd87-43ca-b941-882f3b91d842)

Macros can also be used (e.g., to pull the minimum or maximum year from an options field or a post field). 

**For instance:**

- On the Options page, there’s a field “start_year” with a value equal to 2020.

![image](https://github.com/user-attachments/assets/5d33497e-9c1e-4984-adf6-daeb3faa3834)

- Generate a macro to retrieve the field value:

![image](https://github.com/user-attachments/assets/bf270b6e-4477-4796-8bff-18f871bdaa57)

- Use it in the field:

![image](https://github.com/user-attachments/assets/634d25ef-50f3-4e9b-9cc8-b9f2a68e9275)

Now, the calendar’s minimum year is pulled from the Options Pages’ field.

## Drag-and-Drop Custom Ordering for Related Items on the Back End with Front-End Result Display

Now, you can drag and drop related items to change their order on the back end, and the changes will be reflected on the front end.

![image](https://github.com/user-attachments/assets/f4621cd7-7b5b-4cb9-8c44-3ddb751bc47c)

The only problem here - each time we need to set new order of items in the DB. If do this on the fly, after item reordering - it will use a lot of server resources and could cause UX problems. So, after reordering items, user need to manually confirm re-ordering and after this the new order will be written into the DB. If user reload page without confirming re-order, these changes will be lost.

![image](https://github.com/user-attachments/assets/c6c4d0ea-d477-4599-8196-9dbb37302f4c)

### How to apply the new order on the front end

After changing the order in the admin panel, all methods (macros, dynamic tags, etc.) that return a list of related IDs for a given post (user, term, etc.) will reflect the new order. However, suppose you are not just displaying this list of IDs directly but passing it as an argument to a query builder. In that case, you need to specify in the sorting settings of the corresponding query to order the results according to the sequence of IDs in the provided option. The settings and sorting options vary for each query type.

**Querying related items for Posts**

- Pass the list of IDs to the “Post In” option.
- Sort by “Preserve post ID order given in the “Post In” option.”

![image](https://github.com/user-attachments/assets/db06e319-6453-4332-aa99-5b5c20330baf)
![image](https://github.com/user-attachments/assets/81722933-dc51-46a0-aa10-65361a99a307)

**Querying related items for Users**

- Pass the list of IDs to the “Include” option.
- Sort by “By the included list of user IDs (requires the Include parameter).”

![image](https://github.com/user-attachments/assets/be3d6054-b3b0-4756-946a-ed66a496e9f2)
![image](https://github.com/user-attachments/assets/a65d19c0-a7f2-4b06-87bb-f718bdf458bc)

**Querying related items for Taxonomy Terms**

- Pass the list of IDs to the “Include” option.
- Sort by “Match the order of the Include param.”

![image](https://github.com/user-attachments/assets/f6dde843-ecfb-4564-9bb9-c034094b00ee)
![image](https://github.com/user-attachments/assets/af123d64-56ce-4c1f-8605-e8a2ebc7b76e)

**For Custom Content Types (CCT)**

- Pass the list of IDs in the Query, with the Field set to “Item ID,” Compare – “In the list,” and the Value field containing a list of related IDs.
- Sort by “Preserve item ID order given in the query arguments.”

![image](https://github.com/user-attachments/assets/9b1e4490-ec78-42e6-a89e-00031e935d76)
![image](https://github.com/user-attachments/assets/67aa259a-b595-4a28-85b6-e29e57e4475c)

## Quick Query Editing for Listings on the Front End

When you hover over a widget/block containing a query (only available for JetEngine widgets), the "Edit Query" button will appear in the top right corner, as shown here: 

<img width="613" alt="image" src="https://github.com/user-attachments/assets/673895b2-524f-4f03-8a55-bb094f1685d9">

After clicking this button, the pop-up opens where you can edit the query like this: 

<img width="628" alt="image" src="https://github.com/user-attachments/assets/6600377f-2df8-48b0-8710-a56a29fc1299">

This button is visible only to administrators.

If you want to disable this feature, add the following code to your theme’s `functions.php` file or into a PHP snippet:

```php
add_filter( 'jet-engine/query-builder/frontend-editor/is-enabled', '__return_false' );
```

## Rich Text Сontrol Type for Components

A new control type, Rich Text, has been added.

![image](https://github.com/user-attachments/assets/0ea14b1d-06b7-4d05-9304-3e9f1e5eba7b)

For each builder where a component instance is used, this control is added through the corresponding control type for that builder. In Bricks and Elementor, these controls appear similarly:

![image](https://github.com/user-attachments/assets/8e677571-9b27-422f-9647-542226d21218)

For the Block Editor, however, things are more complex, as this type of control is only supported within the main editor canvas, not in the sidebar. To accommodate this, a custom solution was created:

![image](https://github.com/user-attachments/assets/020bf71f-d020-4576-b362-28e1a17ec3eb)

For sidebar Rich Text controls, only an "Edit HTML" button is available. When clicked, the component content in the editor canvas is replaced with the control-specific content for advanced editing. Once you click "Done," the main component returns with the applied control settings.

![image](https://github.com/user-attachments/assets/e70d1581-7ad5-4c7d-a505-97511793867c)

## Timber/Twig Editor. Emmet support

HTML and CSS editors (if you use the Timber/Twig View) now support the Emmet syntax - https://docs.emmet.io/abbreviations/. Abbreviations turn into code if you click on the tab.

![image](https://github.com/user-attachments/assets/453a4b2b-b9a3-4534-a4a4-881a66aa9456)

**Please leave your feedback and any found bugs or issues in the comments below**

**Comment by tobisalami** [comment]
Can't wait to take it for a spin. I think it'll be a game changer, especially for those who already understand how the structure is meant to be. How do we improve the AI feedback? Is there a way we can further train it? 

Once I get a hang of this, I'm going to make a video concerning it.

**Comment by MjHead** [comment]
@tobisalami at the moment we're using completion API to get the results, so you can improve result only by writing new, more specific prompts. Maybe in the future we'll try to migrate to an Assistant API to keep the context while working on the same website

**Comment by amy-turner** [comment]
Regarding the AI Website Structure Builder, this is a solid start on what could be an extremely useful tool.

In my first test, it's not accepting my prompts. When I click the "Preview Website Model" button, it keeps giving the error "Prompt is too long. Limit is 450 chars" no matter how many characters I have entered. I can't get past that point.

Let me know if any other info would be helpful.

![CleanShot 2024-11-10 at 12 09 55@2x](https://github.com/user-attachments/assets/31b4344c-fe9d-4012-9f14-c44878fc8714)

**Comment by MjHead** [comment]
@amy-turner Hi! Thanks for the feedback, we already fixed this issue and tomorrow will upload 3.6.0-beta2 to the account

**Comment by yasirshabbirservices** [comment]
I just watched Paul video and it looks just game changer :)
